### PR TITLE
[IDEA-37227] Add new "Boolean expression may be factorized" inspection

### PR DESCRIPTION
**Pull request by Fabrice-TIERCELIN**
[Creates a new inspection that factorizes boolean expression when possible](https://youtrack.jetbrains.com/issue/IDEA-37227):

`isValid && isFound || isValid && isActive`

becomes:

`isValid && (isFound || isActive)`

### CODE REVIEW

**Comment by Fabrice-TIERCELIN** [comment]
@BasLeijdekkers,

I have managed to squash the Pull Request of the #1550 but I deleted and recreated my repository to do so. The unexpected result is that the diff is not updated so I recreate the PR.

**Comment by Fabrice-TIERCELIN** [comment]
@BasLeijdekkers,

Yes, it can be applied to numeric too. I wanted to code a simple dev first and extend it later.
But you may think we should not rebrand this inspection in the future...
Do you prefer me to code this inspection for numeric now?

File: plugins/InspectionGadgets/InspectionGadgetsAnalysis/resources/messages/InspectionGadgetsBundle.properties [hunk 1]
```diff
@@ -1658,6 +1658,9 @@ variable.not.used.inside.conditional.problem.descriptor=<code>#ref</code> checke
 boolean.expression.may.be.conditional.display.name=Boolean expression could be replaced with conditional expression
 if.may.be.conditional.problem.descriptor=<code>#ref</code> could be replaced with conditional expression #loc
 if.may.be.conditional.quickfix=Replace with conditional expression
+boolean.expression.may.be.factorized.display.name=Boolean expression could have a factorized expression
```
Comment (BasLeijdekkers) [inline plugins/InspectionGadgets/InspectionGadgetsAnalysis/resources/messages/InspectionGadgetsBundle.properties:1661]: I think an inspection name like "Boolean expression can be factorized" would be shorter and clearer.

File: plugins/InspectionGadgets/InspectionGadgetsAnalysis/resources/messages/InspectionGadgetsBundle.properties [hunk 2]
```diff
@@ -1658,6 +1658,9 @@ variable.not.used.inside.conditional.problem.descriptor=<code>#ref</code> checke
...
+if.may.be.factorized.problem.descriptor=<code>#ref</code> could have a factorized expression #loc
```
Comment (BasLeijdekkers) [inline plugins/InspectionGadgets/InspectionGadgetsAnalysis/resources/messages/InspectionGadgetsBundle.properties:1662]: For the problem descriptor as well: "x an be factorized"

File: plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/controlflow/BooleanExpressionMayBeFactorizedInspection.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      final PsiBinaryExpression lBinaryExpression = (PsiBinaryExpression)lhs;
+      final PsiBinaryExpression rBinaryExpression = (PsiBinaryExpression)rhs;
+      final IElementType lTokenType = lBinaryExpression.getOperationTokenType();
+      final IElementType rTokenType = rBinaryExpression.getOperationTokenType();
+      if (!JavaTokenType.ANDAND.equals(lTokenType) || !JavaTokenType.ANDAND.equals(rTokenType)) {
+        return;
+      }
+      final PsiExpression expression1 = PsiUtil.skipParenthesizedExprDown(lBinaryExpression.getLOperand());
+      final PsiExpression expression2 = PsiUtil.skipParenthesizedExprDown(rBinaryExpression.getLOperand());
+      if (EquivalenceChecker.getCanonicalPsiEquivalence().expressionsAreEquivalent(expression1, expression2) && !SideEffectChecker.mayHaveSideEffects(expression1)) {
```
Comment (BasLeijdekkers) [inline plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/controlflow/BooleanExpressionMayBeFactorizedInspection.java:130]: NPE on `() && b || () && c`.

File: plugins/InspectionGadgets/src/META-INF/InspectionGadgets.xml [hunk 4]
```diff
@@ -578,6 +578,10 @@
                      key="boolean.expression.may.be.conditional.display.name" groupBundle="messages.InspectionsBundle"
                      groupKey="group.names.control.flow.issues" enabledByDefault="false" level="WARNING"
                      implementationClass="com.siyeh.ig.controlflow.BooleanExpressionMayBeConditionalInspection" cleanupTool="true"/>
+    <localInspection groupPath="Java" language="JAVA" shortName="BooleanExpressionMayBeFactorized" bundle="messages.InspectionGadgetsBundle"
+                     key="boolean.expression.may.be.factorized.display.name" groupBundle="messages.InspectionsBundle"
+                     groupKey="group.names.control.flow.issues" enabledByDefault="false" level="WARNING"
```
Comment (BasLeijdekkers) [inline plugins/InspectionGadgets/src/META-INF/InspectionGadgets.xml:583]: Please enable by default with `level="INFORMATION"`.

File: plugins/InspectionGadgets/src/inspectionDescriptions/BooleanExpressionMayBeFactorized.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<html>
+<body>
+Reports any <code>boolean</code> expressions used twice that can be factorized, in order to reduce redundancy and improve the readability.
+Take for example the following expression:
+<pre>a &amp;&amp; b || a &amp;&amp; c;</pre>
```
Comment (BasLeijdekkers) [inline plugins/InspectionGadgets/src/inspectionDescriptions/BooleanExpressionMayBeFactorized.html:5]: For the 2021.2 release we have standardized the format of the example part of the description. It looks like this:
```
<p><b>Example:</b></p>
<pre><code>
  <!-- example here -->
</code></pre>
<p>After the quick-fix is applied:</p>
<pre><code>
  <!-- fixed example here -->
</code></pre>
```
Also you could remove the semicolons. They are not relevant for the example and the statement is not compilable anyway.

**Comment by BasLeijdekkers** [comment]
It would be nice to be able to factorize numeric expressions as well at some point, but I think it is not necessary for this pull request. However it seems to me the following boolean can also be factorized:
```
boolean w = (a || b) && (a || c);
```
to:
```
boolean w = a || b && c;
```
Do you think the inspection can be modified to handle this as well?

**Review by BasLeijdekkers — CHANGES_REQUESTED** [review]

**Pull request closed without merging**